tunnel/intercept/svcpoll.go

```diff
@@ -21,58 +21,86 @@ import (
 	"github.com/netfoundry/ziti-edge/tunnel/dns"
 	"github.com/netfoundry/ziti-sdk-golang/ziti"
 	"github.com/netfoundry/ziti-sdk-golang/ziti/edge"
-	"github.com/sirupsen/logrus"
 	"io"
 	"net"
+	"os"
+	"os/signal"
 	"strings"
+	"syscall"
 	"time"
 )
 
 func ServicePoller(context ziti.Context, interceptor Interceptor, resolver dns.Resolver, pollRate time.Duration) {
 	interceptor.Start(context)
 	knownServices := make(map[string]edge.Service)
+	log := pfxlog.Logger()
+
+	sig := make(chan os.Signal)
+	signal.Notify(sig)
 
 	if pollRate < time.Second {
 		pollRate = 15 * time.Second
 	}
 
-	for ; ; time.Sleep(pollRate) {
+	for {
 		edgeServices, err := context.GetServices()
 		if err != nil {
-			logrus.Errorf("failed to get ziti services: %v", err)
+			log.Errorf("failed to get ziti services: %v", err)
 			if err.Error() == "unauthorized" {
 				if err := context.Authenticate(); err != nil {
-					logrus.WithError(err).Error("could not re-authenticate, session lost")
-					return
+					log.WithError(err).Error("could not re-authenticate, session lost")
+					break
 				}
 			}
-
-			continue
 		}
+		added, removed := diffServices(edgeServices, knownServices)
+		updateServices(context, interceptor, resolver, added, removed, knownServices)
 
-		// find new services
-		addedServices := make(map[string]edge.Service)
-		edgeServiceIds := make(map[string]struct{})
-		for _, edgeSvc := range edgeServices {
-			// get edge service IDs for efficiently finding removed services
-			edgeServiceIds[edgeSvc.Id] = struct{}{}
-			if _, ok := knownServices[edgeSvc.Id]; !ok {
-				addedServices[edgeSvc.Id] = edgeSvc
-				knownServices[edgeSvc.Id] = edgeSvc
+		select {
+		case <-time.After(pollRate):
+			continue
+		case s := <-sig:
+			log.Debugf("caught signal %v", s)
+			switch s {
+			case os.Interrupt, os.Kill, syscall.SIGTERM:
```

Comment:
any reason for variance in namespace? Also what about other terminating signals (QUIT, ABORT, SEGV)?

also can we handle suspend/wakeup (SIGSTOP/SIGCONT) sequence?

Reply:
At one point we were only using the more portable `os` package, which only deals with signet and sigkill. Eventually sigterm was added, and I should've changed the others to `syscall` at that time. Done now.

As for stop/cont, we could do that but I'm not 100% clear on how useful it is?
- SIGSTOP would probably remove all intercepts but remain in the loop
- SIGCONT would probably `continue` the loop to ensure a quick poll for services after waking up (regardless of the poll rate)

Reply:
the process could be suspended and then terminated without giving you the chance to clear the routes. It is probably rare set of circumstances but it seems that it won't be hard to handle for completeness